File: ArraySimulation/MPPT/GlobalMPPTAlgorithms/ParticleSwarm.py
"""
ParticleSwarm.py

Author: Afnan Mir, Array Lead (2021).
Contact: afnanmir@utexas.edu
Created: 08/08/2021
Last Modified: 08/08/2021

Description: Implementation of Particle Swarm Optimization GlobalMPPTAlgortihm.
"""

#library imports
from hashlib import new
import random
import math

#custom imports
from ArraySimulation.MPPT.GlobalMPPTAlgorithms.GlobalMPPTAlgorithm import (
    GlobalMPPTAlgorithm,
)

class Particle():
    def __init__(self, xPos, vel):
        self.personalBest = 0.0
        self.personalBestVolt = 0.0
        self.xPos = xPos
        self.vel = vel
    
    def changeXPos(self,vel):
        self.xPos += vel
    
    def changeVel(self, newVel):
        self.vel = newVel
    
    def getPBest(self):
        return self.personalBest

    def getVoltBest(self):
        return self.personalBestVolt

    def getXPos(self):
        return self.xPos
    
    def getVel(self):
        return self.vel

    def changePBest(self, newBest, newBestVolt):
        self.personalBest = newBest
        self.personalBestVolt = newBestVolt
    def __str__(self):
        return f"Position: {self.xPos}\n Velocity: {self.vel}\nPersonal Best: {self.personalBestVolt}"


class ParticleSwarm(GlobalMPPTAlgorithm):
    """
    Class to implement the Particle Swarm Optimization Algorithm for Maximium Power Point Tracking
    (https://www.ieeexplore-ieee-org.ezproxy.lib.texas.edu/document/7233061)
    """
    NUM_AGENTS = 4
    C1 = 0.02
    C2 = 0.5
    W = 0.4

    def __init__(self, 
        numCells=1, 
        MPPTLocalAlgoType="Default",
        strideType="Fixed"):
        """
        Constructor for the Particle Swarm Optimization algorithm

        Parameters
        ----------
        numCells : int
            number of cells
        MPPTLocalAlgoType : str
            the local algorithm to use
        strideType : str
            Stride to use for perturb and observe

        Returns
        -------
        None
        """
        super(ParticleSwarm, self).__init__(numCells, "Particle Swarm", MPPTLocalAlgoType, strideType)
        self.gBest = 0.0
        self.gBestVolt = 0.0
        self.agents = []
        self.goForward = True
        self._setup = True
        self.startLocal = True
        self.kick = True
        self.totalCycle = 0
        interval = GlobalMPPTAlgorithm.MAX_VOLTAGE/5
        for i in range(ParticleSwarm.NUM_AGENTS):
            self.agents.append(Particle((random.random()*interval)+(interval*i), 0.0))
        self.cycle = -1
        for i in self.agents:
            print(i.getXPos())


    def getVelocityVector(self, agent):
        """
        Returns the new velocity for the agent given the agents personal best position, the global best position, and the current velocity based
        on the equation
        V_{k+1} = w*V_{k} + c1*r1*(p_{i_best} - x_{k}) + c2*r2*(g_best - x_{k})
        V_k = current velocity
        c1 and c2: constants
        r1 and r2: normalized random number between 0 and 1.
        p_{i_best} = current best position the agent has been to
        g_best = global best position
        w = constant

        Parameters
        ----------
        agent : Particle
            The particle that is being updated

        Returns
        -------
        newVel : int
            the new velocity vector for the agent.
        """
        r1 = random.random()
        r2 = random.random()
        newVel = ParticleSwarm.W*agent.getVel() + ParticleSwarm.C1*r1*(agent.getVoltBest() - agent.getXPos()) + ParticleSwarm.C2*r2*(self.gBestVolt - agent.getXPos())
        # print(newVel)
        return newVel

    def setUp(self, arrVoltage, arrCurrent, irradiance, temperature):
        """
        Method takes care of the first 4 cycles to put each agent in its starting point and
        determines the personal bests and global best position(s).

        Parameters
        ----------
        arrVoltage : float
            Array voltage of the current cycle
        arrCurrent : float
            Array current of the current cycle
        irradiance : float
            Irradiance of the current cycle
        temperature : float
            Temperature of the current cycle

        Returns
        -------
        vRef : float
            next array voltage to output
        """
        agentPower = arrVoltage * arrCurrent
        if(self.cycle == -1):
            self.cycle = 0
            self.totalCycle += 1
            vRef =  self.agents[0].getXPos()
        elif(self.cycle == 0 and self._setup):
            if(agentPower > self.agents[0].getPBest()):
                self.agents[0].changePBest(agentPower, arrVoltage)
            if(agentPower > self.gBest):
                self.gBest = agentPower
                self.gBestVolt = arrVoltage
            self.cycle += 1
            self.totalCycle += 1
            vRef = self.agents[1].getXPos()
        elif(self.cycle == 1 and self._setup):
            if(agentPower > self.agents[1].getPBest()):
                self.agents[1].changePBest(agentPower, arrVoltage)
            if(agentPower > self.gBest):
                self.gBest = agentPower
                self.gBestVolt = arrVoltage
            self.totalCycle += 1
            self.cycle += 1
            vRef = self.agents[2].getXPos()
        elif(self.cycle == 2 and self._setup):
            if(agentPower > self.agents[2].getPBest()):
                self.agents[2].changePBest(agentPower, arrVoltage)
            if(agentPower > self.gBest):
                self.gBest = agentPower
                self.gBestVolt = arrVoltage
            self.cycle += 1
            self.totalCycle += 1
            vRef = self.agents[3].getXPos()
        elif(self.cycle == 3 and self._setup):
            if(agentPower > self.agents[3].getPBest()):
                self.agents[3].changePBest(agentPower, arrVoltage)
            if(agentPower > self.gBest):
                self.gBest = agentPower
                self.gBestVolt = arrVoltage
            self.totalCycle += 1
            self.cycle = 2
            self._setup = False
            self.goForward = False
            vRef = self.agents[2].getXPos()
        return vRef
    def agentUpdate(self, arrVoltage, arrCurrent, irradiance, temperature):
        """
        Method that updates the current state of the particles on the curve.

        Parameters
        ----------
        arrVoltage : float
            Array voltage of the current cycle
        arrCurrent : float
            Array current of the current cycle
        irradiance : float
            Irradiance of the current cycle
        temperature : float
            Temperature of the current cycle

        Returns
        -------
        vRef : float
            array voltage to output
        """
        vRef = arrVoltage
        if(self._setup):
            vRef = self.setUp(arrVoltage, arrCurrent, irradiance, temperature)
        else:
            vRef = arrVoltage
            agentPower = arrVoltage * arrCurrent
            if(self.cycle == 0):
                if(agentPower > self.agents[0].getPBest()):
                    self.agents[0].changePBest(agentPower, arrVoltage)
                if(agentPower > self.gBest):
                    self.gBest = agentPower
                    self.gBestVolt = arrVoltage
                newVel = self.getVelocityVector(self.agents[0])
                self.agents[0].changeXPos(newVel)
                self.agents[0].changeVel(newVel)
                self.goForward = True
                self.cycle = 1
                self.totalCycle += 1
                vRef = self.agents[1].getXPos()
            elif(self.cycle % 4 == 1):
                if(agentPower > self.agents[1].getPBest()):
                    self.agents[1].changePBest(agentPower, arrVoltage)
                if(agentPower > self.gBest):
                    self.gBest = agentPower
                    self.gBestVolt = arrVoltage
                newVel = self.getVelocityVector(self.agents[1])
                self.agents[1].changeXPos(newVel)
                self.agents[1].changeVel(newVel)
                if(self.goForward):
                    vRef = self.agents[2].getXPos()
                    self.cycle = 2
                else:
                    vRef = self.agents[0].getXPos()
                    self.cycle = 0
                self.totalCycle += 1
            elif(self.cycle % 4 == 2):
                if(agentPower > self.agents[2].getPBest()):
                    self.agents[2].changePBest(agentPower, arrVoltage)
                if(agentPower > self.gBest):
                    self.gBest = agentPower
                    self.gBestVolt = arrVoltage
                newVel = self.getVelocityVector(self.agents[2])
                self.agents[2].changeXPos(newVel)
                self.agents[2].changeVel(newVel)
                if(self.goForward):
                    vRef = self.agents[3].getXPos()
                    self.cycle = 3
                else:
                    vRef = self.agents[1].getXPos()
                    self.cycle = 1
                self.totalCycle += 1
            elif(self.cycle % 4 == 3):
                if(agentPower > self.agents[3].getPBest()):
                    self.agents[3].changePBest(agentPower, arrVoltage)
                if(agentPower > self.gBest):
                    self.gBest = agentPower
                    self.gBestVolt = arrVoltage
                newVel = self.getVelocityVector(self.agents[3])
                self.agents[3].changeXPos(newVel)
                self.agents[3].changeVel(newVel)
                vRef = self.agents[2].getXPos()
                self.cycle = 2
                self.totalCycle += 1
                self.goForward = False
        return vRef

    def getReferenceVoltage(self, arrVoltage, arrCurrent, irradiance, temperature):
        """
        Method that updates output voltage of the array.

        Parameters
        ----------
        arrVoltage : float
            Array voltage of the current cycle
        arrCurrent : float
            Array current of the current cycle
        irradiance : float
            Irradiance of the current cycle
        temperature : float
            Temperature of the current cycle

        Returns
        -------
        vRef : float
            array voltage to output
        """
        if(self.totalCycle <= 45):
            vRef =  self.agentUpdate(arrVoltage,arrVoltage, irradiance, temperature)
            print(f"Cycle {self.totalCycle}")
            for i in range(len(self.agents)):
                print("Particle" + str(i))
                print(self.agents[i])
        else:
            if(self.startLocal):
                vRef = self.gBestVolt
                self.startLocal = False #start converging to global maximum
            elif(self.kick): #start local mppt algorithm.
                vRef = arrVoltage + 0.02
                self.kick = False
                self.vOld =arrVoltage
                self.pOld = arrCurrent * arrVoltage
                self._model._strideModel.vOld = arrVoltage
                self._model._strideModel.pOld = self.pOld
                self.iOld = arrCurrent
            else:
                vRef = self._model.getReferenceVoltage(
                    arrVoltage, arrCurrent, irradiance, temperature
                )
            self.totalCycle += 1
            return vRef
        return vRef
    def reset(self):
        """
        Method to reset the pipeline

        Parameters
        ----------
        None.

        Returns
        -------
        None.
        """
        self.gBest = 0.0
        self.gBestVolt = 0.0
        # self.agents = []
        self.goForward = True
        self._setup = True
        self.startLocal = True
        self.kick = True
        self.totalCycle = 0
        interval = GlobalMPPTAlgorithm.MAX_VOLTAGE/5
        for i in range(ParticleSwarm.NUM_AGENTS):
            self.agents.append(Particle((random.random()*interval)+(interval*i), 0.0))
        self.cycle = -1
        for i in self.agents:
            print(i.getXPos())
        

File: ArraySimulation/MPPT/GlobalMPPTAlgorithms/ParticleSwarm.cpp
//
// Created by mpran on 5/22/2023.
//

#include "ParticleSwarm.h"

Particle::Particle(float xPos, float vel) : xPos(xPos), vel(vel) {
    this->personalBest = 0.0f;
    this->personalBestVolt = 0.0f;
}

void Particle::changeXPos(float vel) {
    this->xPos += vel;
}

void Particle::changeVel(float newVel) {
    this->vel = newVel;
}

float Particle::getPBest() {
    return this->personalBest;
}

float Particle::getVoltBest() {
    return this->personalBestVolt;
}

float Particle::getXPos() {
    return this->xPos;
}

float Particle::getVel() {
    return this->vel;
}

void Particle::changePBest(float newBest, float newBestVolt) {
    this->personalBest = newBest;
    this->personalBestVolt = newBestVolt;
}

string Particle::toString() {
    string result = "Position: " + to_string(this->xPos) + "\nVelocity: " + to_string(this->vel) + "\nPersonal Best: " + to_string(this->personalBestVolt);
    return result;
}

ParticleSwarm::ParticleSwarm(int numCells, string MPPTLocalAlgoType, string strideType) : GlobalMPPTAlgorithm(numCells, "Particle Swarm", MPPTLocalAlgoType, strideType){
    gBest = 0;
    gBestVolt = 0;
    goForward = true;
    _setup = true;
    startLocal = true;
    kick = true;
    totalCycle = 0;
    float interval = GlobalMPPTAlgorithm::MAX_VOLTAGE / 5;
    for(int i = 0; i < ParticleSwarm::NUM_AGENTS; ++i) {
        float xPos = (static_cast<float>(rand()) / RAND_MAX) * interval + interval * i;
        agents.push_back(new Particle(xPos, 0.0f));
    }
}

float ParticleSwarm::getVelocityVector(Particle* agent) {
    float r1 = static_cast<float>(rand()) / RAND_MAX;
    float r2 = static_cast<float>(rand()) / RAND_MAX;

    float newVel = ParticleSwarm::W * agent->getVel() +
                   ParticleSwarm::C1 * r1 * (agent->getVoltBest() - agent->getXPos()) +
                   ParticleSwarm::C2 * r2 * (this->gBestVolt - agent->getXPos());

    return newVel;
}

float ParticleSwarm::setUp(float arrVoltage, float arrCurrent, float irradiance, float temperature) {
    float agentPower = arrVoltage * arrCurrent;
    float vRef;

    if(this->cycle == -1) {
        this->cycle = 0;
        this->totalCycle += 1;
        vRef = this->agents[0]->getXPos();
    } else if(this->cycle == 0 && this->_setup) {
        if(agentPower > this->agents[0]->getPBest()) {
            this->agents[0]->changePBest(agentPower, arrVoltage);
        }
        if(agentPower > this->gBest) {
            this->gBest = agentPower;
            this->gBestVolt = arrVoltage;
        }
        this->cycle += 1;
        this->totalCycle += 1;
        vRef = this->agents[1]->getXPos();
    } else if(this->cycle == 1 && this->_setup) {
        if(agentPower > this->agents[1]->getPBest()) {
            this->agents[1]->changePBest(agentPower, arrVoltage);
        }
        if(agentPower > this->gBest) {
            this->gBest = agentPower;
            this->gBestVolt = arrVoltage;
        }
        this->totalCycle += 1;
        this->cycle += 1;
        vRef = this->agents[2]->getXPos();
    } else if(this->cycle == 2 && this->_setup) {
        if(agentPower > this->agents[2]->getPBest()) {
            this->agents[2]->changePBest(agentPower, arrVoltage);
        }
        if(agentPower > this->gBest) {
            this->gBest = agentPower;
            this->gBestVolt = arrVoltage;
        }
        this->cycle += 1;
        this->totalCycle += 1;
        vRef = this->agents[3]->getXPos();
    } else if(this->cycle == 3 && this->_setup) {
        if(agentPower > this->agents[3]->getPBest()) {
            this->agents[3]->changePBest(agentPower, arrVoltage);
        }
        if(agentPower > this->gBest) {
            this->gBest = agentPower;
            this->gBestVolt = arrVoltage;
        }
        this->totalCycle += 1;
        this->cycle = 2;
        this->_setup = false;
        this->goForward = false;
        vRef = this->agents[2]->getXPos();
    }

    return vRef;
}

float ParticleSwarm::agentUpdate(float arrVoltage, float arrCurrent, float irradiance, float temperature) {
    float vRef = arrVoltage;
    if (this->_setup) {
        vRef = this->setUp(arrVoltage, arrCurrent, irradiance, temperature);
    } else {
        float agentPower = arrVoltage * arrCurrent;
        float newVel;
        if (this->cycle == 0) {
            if (agentPower > this->agents[0]->getPBest()) {
                this->agents[0]->changePBest(agentPower, arrVoltage);
            }
            if (agentPower > this->gBest) {
                this->gBest = agentPower;
                this->gBestVolt = arrVoltage;
            }
            newVel = this->getVelocityVector(this->agents[0]);
            this->agents[0]->changeXPos(newVel);
            this->agents[0]->changeVel(newVel);
            this->goForward = true;
            this->cycle = 1;
            this->totalCycle++;
            vRef = this->agents[1]->getXPos();
        } else if (this->cycle % 4 == 1) {
            if (agentPower > this->agents[1]->getPBest()) {
                this->agents[1]->changePBest(agentPower, arrVoltage);
            }
            if (agentPower > this->gBest) {
                this->gBest = agentPower;
                this->gBestVolt = arrVoltage;
            }
            newVel = this->getVelocityVector(this->agents[1]);
            this->agents[1]->changeXPos(newVel);
            this->agents[1]->changeVel(newVel);
            if (this->goForward) {
                vRef = this->agents[2]->getXPos();
                this->cycle = 2;
            } else {
                vRef = this->agents[0]->getXPos();
                this->cycle = 0;
            }
            this->totalCycle++;
        } else if (this->cycle % 4 == 2) {
            if (agentPower > this->agents[2]->getPBest()) {
                this->agents[2]->changePBest(agentPower, arrVoltage);
            }
            if (agentPower > this->gBest) {
                this->gBest = agentPower;
                this->gBestVolt = arrVoltage;
            }
            newVel = this->getVelocityVector(this->agents[2]);
            this->agents[2]->changeXPos(newVel);
            this->agents[2]->changeVel(newVel);
            if (this->goForward) {
                vRef = this->agents[3]->getXPos();
                this->cycle = 3;
            } else {
                vRef = this->agents[1]->getXPos();
                this->cycle = 1;
            }
            this->totalCycle++;
        } else if (this->cycle % 4 == 3) {
            if (agentPower > this->agents[3]->getPBest()) {
                this->agents[3]->changePBest(agentPower, arrVoltage);
            }
            if (agentPower > this->gBest) {
                this->gBest = agentPower;
                this->gBestVolt = arrVoltage;
            }
            newVel = this->getVelocityVector(this->agents[3]);
            this->agents[3]->changeXPos(newVel);
            this->agents[3]->changeVel(newVel);
            vRef = this->agents[2]->getXPos();
            this->cycle = 2;
            this->totalCycle++;
            this->goForward = false;
        }
    }
    return vRef;
}

float ParticleSwarm::getReferenceVoltage(float arrVoltage, float arrCurrent, float irradiance, float temperature) {
    float vRef;
    if (this->totalCycle <= 45) {
        vRef = this->agentUpdate(arrVoltage, arrCurrent, irradiance, temperature);
    } else {
        if (this->startLocal) {
            this->model.setup(this->gBestVolt, 0, GlobalMPPTAlgorithm::MAX_VOLTAGE);
            vRef = this->gBestVolt;
            this->startLocal = false;
        } else if (this->kick) {
            vRef = arrVoltage + 0.02;
            this->kick = false;
            this->vOld = arrVoltage;
            this->pOld = arrCurrent * arrVoltage;
            this->model.strideModel.vOld = arrVoltage;
            this->model.strideModel.pOld = this->pOld;
            this->iOld = arrCurrent;
        } else {
            vRef = this->model.getReferenceVoltage(arrVoltage, arrCurrent, irradiance, temperature);
            bool needsChange = this->checkEnvironmentalChanges(irradiance);
            if (needsChange) {
                this->gBest = 0.0;
                this->gBestVolt = 0.0;
                this->agents.clear();
                this->runningHistory.clear();
                this->goForward = true;
                this->_setup = true;
                this->startLocal = true;
                this->kick = true;
                this->totalCycle = 0;
                float interval = GlobalMPPTAlgorithm::MAX_VOLTAGE / 5;
                for (int i = 0; i < ParticleSwarm::NUM_AGENTS; i++) {
                    this->agents.push_back(new Particle((rand() / (float)RAND_MAX * interval) + (interval * i), 0.0));
                }
                this->cycle = -1;
            }
            this->totalCycle++;
        }
    }
    return vRef;
}

ParticleSwarm::~ParticleSwarm() {
    for(auto agent : agents) {
        delete agent;
    }
}

void ParticleSwarm::reset() {
    this->gBest = 0.0;
    this->gBestVolt = 0.0;
    this->goForward = true;
    this->_setup = true;
    this->startLocal = true;
    this->kick = true;
    this->totalCycle = 0;
    float interval = GlobalMPPTAlgorithm::MAX_VOLTAGE / 5;
    this->agents.clear();
    for (int i = 0; i < ParticleSwarm::NUM_AGENTS; i++) {
        this->agents.push_back(new Particle((rand() / (float)RAND_MAX * interval) + (interval * i), 0.0));
    }
    this->cycle = -1;
    for (Particle* p : this->agents) {
        std::cout << p->getXPos() << std::endl;
    }
}